### DISCUSSION TITLE
How to Swizzle Document Datasource?

### DISCUSSION THREAD

**Discussion by lukedukeus** [facebook/docusaurus#8028] [Q&A] [answered]
I am trying to implement route protection, where some documents are public, and other documents are limited to specific users. I have implemented the sign in, and I am setting the scope of the document in its frontmatter:
```
---
sidebar_position: 1
sidebar_custom_props: { scope: public } // or a users name if it is private to them.
---
```

Then, I am wrapping each component that would render documents, to check if that user is logged in.

However, rather than doing this for each component, is it possible to do this at the root? Or can I modify the dataset of docs before it goes out to be rendered, rather than intercepting it in the props?

**Comment by slorber** [comment, marked as answer]
> However, rather than doing this for each component, is it possible to do this at the root?

The root /layout doesn't currently have access frontmatter. Not all pages are Markdown documents with available frontMatter.

You can try to swizzle `MDXContent` and use `useFrontMatter` (from https://github.com/roydukkey/docusaurus-theme-frontmatter, but we'll likely add something like this in core later)

> I am trying to implement route protection, where some documents are public, and other documents are limited to specific users.
> Then, I am wrapping each component that would render documents, to check if that user is logged in.
 
Not sure what your "wrapper protection" is exactly, but keep in mind that JS bundles containing potentially sensitive information are still available on `build/assets/js` as part of your static deployment. Although your UI doesn't present that sensitive information, it remains possible to access it.

Maybe you could also be interested in this upcoming "unlisted" feature? https://github.com/facebook/docusaurus/pull/8004

If you have doc 1 2 3, and if doc 2 is only available for logged-in users, maybe you'll want to make sure the doc navigation navigates from doc 1 to doc 3 instead of trying to navigate to the inaccessible doc2?

> Or can I modify the dataset of docs before it goes out to be rendered, rather than intercepting it in the props?

You can extend the docs plugin, modify created routes/props and whatever you want. However it might not be very easy. See also https://github.com/facebook/docusaurus/issues/4138

--- 

Related auth comment: https://github.com/facebook/docusaurus/issues/958#issuecomment-1057075925

Note with Cloudflare workers and other edge functions, you can do some quite advanced things. For example, you could build 2 versions of your static website (one for anonymous, one for logged-in users). The edge fn can check auth and serve one or the other according to the auth status of the user.

**Reply by lukedukeus** [reply]
Thanks for the response. I hadn't thought about the fact that all docs are still in the client sources. I'm going ahead with building one copy of the site per scope. I'm hosting locally, so I'll just add a server that returns a different copy of the site based on which user is signed in. Another benefit of doing things this way is that I don't have to mess with how search works.
